### PR TITLE
API endpoints for the series deletion API using block storage

### PR DESCRIPTION
**Pull request by ilangofman**
**What this PR does**:
This PR implements the API endpoints for the time series deletion API for block storage. Based on the proposal for the series deletion API [here](https://cortexmetrics.io/docs/proposals/block-storage-time-series-deletion/).

The work done in this PR includes creating, getting and canceling deletion requests. 

- When a delete request is received, a tombstone is created to track the deletion information and state. 
- Getting the delete request information requires reading the tombstone files.
- Cancelling a deletion is done by deleting the tombstone. The cancellation can only be done while the `-purger.delete-request-cancel-period` hasn't passed since the delete request was originally made. 

For the API to be enabled, set `--purger.enable` to `true`.

The new endpoints are located within the purger, where the existing tenant deletion API is also located. However, the files are under a `chunks` subdirectory but contain existing APIs for blocks storage. A future PR will be required to move the purger service out of the `chunks` subdirectory. 

To make it easier to review and get incremental feedback, the implementation of the proposal is going to be split among roughly 4 PR's.
- API endpoints (this PR) 
- Permanent deletion processing
- Query time filtering using tombstones before data is deleted
- Cache invalidation 

Please let me know if you have any questions or concerns. Any feedback is appreciated. Thank you for taking a look at this PR and helping out! 

**Which issue(s) this PR fixes**:
Initial PR of multiple to address #4267

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/cortex/modules.go [hunk 1]
```diff
@@ -801,18 +802,18 @@ func (t *Cortex) initChunksPurger() (services.Service, error) {
...
 
 	// t.RulerStorage can be nil when running in single-binary mode, and rule storage is not configured.
-	tenantDeletionAPI, err := purger.NewTenantDeletionAPI(t.Cfg.BlocksStorage, t.Overrides, util_log.Logger, prometheus.DefaultRegisterer)
+	blockPurger, err := purger.NewBlocksPurgerAPI(t.Cfg.BlocksStorage, t.Overrides, util_log.Logger, prometheus.DefaultRegisterer, t.Cfg.PurgerConfig.DeleteRequestCancelPeriod)
 	if err != nil {
 		return nil, err
 	}
 
-	t.API.RegisterTenantDeletion(tenantDeletionAPI)
+	t.API.RegisterBlocksPurger(blockPurger, t.Cfg.PurgerConfig.Enable)
```
Comment (harry671003) [inline pkg/cortex/modules.go:816]: This flag could be renamed to a more friendly name: like `EnableBlocksSeriesDeletion`?

**Comment by jeromeinsf** [comment]
Can you please also review this PR @alolita @Aneurysm9

**Review by jeromeinsf — APPROVED** [review]

**Comment by ilangofman** [comment]
> I'm a bit confused why files disappeared like `pkg/chunk/purger/tenant_deletion_api.go` and the why the blocks purger is under `pkg/chunk`.
> 
> Did any functionality disappear? 

@bboreham No functionality disappeared. The `pkg/chunk/purger/tenant_deletion_api.go` file was the blocks implementation of the tenant deletion endpoints.  Since I created more endpoints for the series deletion with blocks storage, I just combined both under one file and called it the blocks purger. So all the previous code in `tenant_deletion_api.go` is now in `pkg/chunk/purger/blocks_purger.go`. I can undo that and create individual files for them if you would like. 

> Could the blocks functionality live under `pkg/storage/tsdb` ?

I agree, it is a bit confusing that some of the blocks functionality is under the chunks folder. I can try moving it to the suggested folder. If the answer to my previous question was to keep the `pkg/chunk/purger/tenant_deletion_api.go` separate, I am guessing it would make sense to move that as well to this folder, since it is only for blocks storage?

File: pkg/cortex/modules.go (hunk 1 above)
Comment (ilangofman) [reply pkg/cortex/modules.go]: I've renamed the variable to `EnableSeriesDeletion`. I didn't change the flag name itself since the flag can be shared among both the chunks and blocks implementation. The current description of the flag is: `Enable purger to allow deletion of series. Be aware that Delete series feature is still experimental`.

**Comment by bboreham** [comment]
> I just combined both under one file and called it the blocks purger

This would be a good thing to explain in the PR description, to aid reviewers.

Also you could organise the change into commits where one is a pure refactor, just moving code, and then subsequent ones modify the behaviour.  The current set of 22 commits shows the journey that you went on, but is not a good breakdown for reviewing.

**Comment by bboreham** [comment]
I expect when we remove chunks support (which is deprecated), then we can delete `pkg/chunks`, so don't add anything blocks-related under there.

**Comment by bboreham** [comment]
Can you explain the distinction between "delete request" and "tombstone" as used in this code?

**Comment by bboreham** [comment]
Adding to my earlier comment, the `TenantDeletionAPI` which only works on blocks was previously added under `pkg/chunk/purger`, so it was already confusing. 

It's really hard to follow when you add functionality and clean old things up at the same time, so I hope what will work is to put the existing code back where it was, then add your new parts and call out to those.  Then we can file a separate issue to migrate all the blocks code out from `pkg/chunk/purger`.

File: pkg/chunk/purger/blocks_purger.go [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+		level.Error(util_log.Logger).Log("msg", "error cancelling the delete request", "err", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.WriteHeader(http.StatusNoContent)
+}
+
+// Calculates the tombstone file name based on a hash of the start time, end time and selectors
+func getTombstoneRequestID(startTime int64, endTime int64, selectors []string) string {
```
Comment (bboreham) [inline pkg/chunk/purger/blocks_purger.go:228]: I prefer to have functions named after what they do, not after how you expect to use them.
So this can be `getTombstoneHash`; the current function with that name doesn't need to exist.

File: pkg/chunk/purger/blocks_purger.go [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+	var b strings.Builder
+	b.WriteString(strconv.FormatInt(startTime, 10))
+	b.WriteString(",")
+	b.WriteString(strconv.FormatInt(endTime, 10))
+
+	sort.Strings(selectors)
+	for _, s := range selectors {
+		b.WriteString(",")
+		b.WriteString(s)
+	}
```
Comment (bboreham) [inline pkg/chunk/purger/blocks_purger.go:239]: Don't make a string, just write the data to a hasher (`hash.Hash` includes `io.Writer`)

File: pkg/storage/tsdb/tombstones.go [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+	logger log.Logger) *TombstoneManager {
+
+	return &TombstoneManager{
+		bkt:    bucket.NewUserBucketClient(userID, bkt, cfgProvider),
+		logger: util_log.WithUserID(userID, logger),
+	}
+}
+
+// Uploads a tombstone file to object sotre
+func (m *TombstoneManager) WriteTombstoneFile(ctx context.Context, tombstone *Tombstone) error {
```
Comment (bboreham) [inline pkg/storage/tsdb/tombstones.go:85]: Is it a "file"?

File: pkg/storage/tsdb/tombstones.go [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+
+	deletionRequests := []*Tombstone{}
+	for _, t := range tombstoneMap {
+		deletionRequests = append(deletionRequests, t)
+	}
+
+	return deletionRequests, nil
+}
+
+func (m *TombstoneManager) getLatestTombstateByState(a *Tombstone, b *Tombstone) (*Tombstone, error) {
```
Comment (bboreham) [inline pkg/storage/tsdb/tombstones.go:183]: Tombstate->Tombstone?

File: pkg/storage/tsdb/tombstones.go [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	orderB, err := b.State.GetStateOrder()
+	if err != nil {
+		return nil, err
+	}
+
+	if orderB > orderA {
```
Comment (bboreham) [inline pkg/storage/tsdb/tombstones.go:194]: If this function is called "latest" shouldn't it be checking the dates?
Good example where a unit test would help clarify what is meant.

**Review by bboreham — COMMENTED** [review]
Looks like this is coming together, but I have a few comments.

File: pkg/chunk/purger/blocks_purger.go (hunk 2 above)
Comment (ilangofman) [reply pkg/chunk/purger/blocks_purger.go:228]: Makes total sense. Will keep that in mind going forward, thanks.

File: pkg/storage/tsdb/tombstones.go (hunk 5 above)
Comment (ilangofman) [reply pkg/storage/tsdb/tombstones.go:183]: Good catch, thanks.

**Comment by ilangofman** [comment]
> Can you explain the distinction between "delete request" and "tombstone" as used in this code?

The way I have been thinking about it is that the tombstone is just a file to keep track of a series deletion request. Getting the delete request information would mean reading the tombstone for it. Admittedly, I have been using the two terms synonymously. I'll update the code to be more consistent. 
If you see any issues with that, please let me know.

File: pkg/storage/tsdb/tombstones.go (hunk 4 above)
Comment (ilangofman) [reply pkg/storage/tsdb/tombstones.go:85]: It does write a file to the object store which is why I originally named it that. However putting "file" might be redundant, just the same way when writing the block deletion marker, it doesn't say "marker file". I've renamed it to just `WriteTombstone`, cause I think the file part is implied. Do you see any potential concerns with this?

File: pkg/storage/tsdb/tombstones.go (hunk 6 above)
Comment (ilangofman) [reply pkg/storage/tsdb/tombstones.go:194]: All the tombstones start with the pending state and can only move to cancelled or processed. Since those are the only ways they can transition, we can get the newest tombstone solely based on the state. 

This function was meant to be comparing 2 tombstones with the same request ID but different states. With that function name, it can be slightly misleading and confusing. To counter this confusion, I've refactored the only function that called this function, and put the logic inside there (`GetAllTombstonesForUser`). Hopefully it is more clear. 

Side note: I also refactored `GetAllTombstonesForUser` so it only reads the tombstone contents after all the duplicate files have been filtered out. This should reduce the total number of calls to the object store. 

I know with the extra refactoring this can cause more confusion when reviewing PR's, sorry about that. 

> Good example where a unit test would help clarify what is meant.

The unit tests for the calling function were testing this functionality. I see how not having it for the individual function could cause confusion. Having them for this function could have made things more clear.

**Comment by jeromeinsf** [comment]
@ilangofman Are you updating the code to be more consistent or should someone take over?

**Comment by ilangofman** [comment]
> @ilangofman Are you updating the code to be more consistent or should someone take over?


Hi @jeromeinsf . I took a couple of weeks off but I am going to get back to this. I think this PR I should be able to finish on my own. 

For the rest of the deletion PR's, I am not sure how much time I will have available to work on them, with university starting this week. I would still like to contribute when I can but it might take some time to make changes. To speed things up, if someone wants to take over the other PR's or collaborate together, I'm open to all!

File: pkg/chunk/purger/blocks_purger.go [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	if err := json.NewEncoder(w).Encode(deleteRequests); err != nil {
+		level.Error(util_log.Logger).Log("msg", "error marshalling response", "err", err)
+		http.Error(w, fmt.Sprintf("Error marshalling response: %v", err), http.StatusInternalServerError)
+	}
+
+	w.WriteHeader(http.StatusOK)
```
Comment (alvinlin123) [inline pkg/chunk/purger/blocks_purger.go:160]: According to https://pkg.go.dev/net/http#ResponseWriter since you  called `json.NewEncoder(w)`, which I assume calls `ResponseWriter.Write()`, then status code of 200 is automatically written for you together with "guessed" content-type. So you don't really need to set the status at this line. 

However, I think we should set the content type and the status header ourselves for clarity. 

Also, I think the if block on line 155 is missing a return statement right? I would expect `http.Error()` writes that the status header  to 500 already, and this `w.WriteHeader(http.StatusOk)` may have undefined behaviour?

File: pkg/chunk/purger/blocks_purger.go [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+func (api *BlocksPurgerAPI) CancelDeleteRequestHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	userID, err := tenant.TenantID(ctx)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusUnauthorized)
+		return
+	}
+
+	params := r.URL.Query()
+	requestID := params.Get("request_id")
```
Comment (alvinlin123) [inline pkg/chunk/purger/blocks_purger.go:173]: Shouldn't we do input validation here? What if  `request_id` is not given?

File: pkg/chunk/purger/blocks_purger.go [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+		return
+	}
+
+	if deleteRequest == nil {
+		http.Error(w, "could not find delete request with given id", http.StatusBadRequest)
+		return
+	}
+
+	if deleteRequest.State == cortex_tsdb.StateCancelled {
+		http.Error(w, "the request has already been previously deleted", http.StatusBadRequest)
```
Comment (alvinlin123) [inline pkg/chunk/purger/blocks_purger.go:189]: In this case, shouldn't we return 202? Because it's already in cancelled state,  which is what this API should do right? 

By the way, in our error messages to the user, we gotta use "cancel" and "delete" carefully. I think Here you actually mean "the series deletion request was cancelled previously" not "the series has already been deleted" right?

File: pkg/chunk/purger/blocks_purger.go [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+		return
+	}
+
+	if deleteRequest.State == cortex_tsdb.StateProcessed {
```
Comment (alvinlin123) [inline pkg/chunk/purger/blocks_purger.go:193]: Do we need a "processing" state? Or we can rely on the `deleteRequestCancelPeriod` to infer the "processing" state?

File: pkg/storage/tsdb/tombstones.go [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+			orderA, err := state.GetStateOrder()
+			if err != nil {
+				return err
+			}
+			orderB, err := prevState.GetStateOrder()
+			if err != nil {
+				return err
+			}
+
+			// If the newer state found comes later in the order, then we replace the state in the map.
```
Comment (alvinlin123) [inline pkg/storage/tsdb/tombstones.go:169]: In this case, I think it might be better to name the variables `newerStateOrder` and `prevStateOrder`; this do imply that if we rename `state` to `newerState` the readability will improve too :)

File: pkg/storage/tsdb/tombstones.go [hunk 12]
```diff
@@ -0,0 +1,360 @@
...
+	case
+		StatePending,
+		StateProcessed,
+		StateCancelled:
+		return true
+	}
+	return false
+}
+
+func (s BlockDeleteRequestState) GetStateOrder() (int, error) {
```
Comment (alvinlin123) [inline pkg/storage/tsdb/tombstones.go:349]: I don't think it's a wise idea to export this method. This looks a lot like implementation detail that should not be exposed. I don't think we want anyone outside the `tsdb` package to depend on the fact that the "order" of `StatePending` is 0, "StateProcessed" is 2, etc. right? Because what is important is the relation between the "order" not the absolute value right? 

If we really want to expose something, consider  functions like `IsGreaterThan(a, b BlockDeleteRequestState)`, but I don't think we should in this case.

File: pkg/storage/tsdb/tombstones.go [hunk 13]
```diff
@@ -0,0 +1,360 @@
...
+	newT.Matchers = t.Matchers
+
+	err := m.WriteTombstone(ctx, newT)
+	if err != nil {
+		level.Error(m.logger).Log("msg", "error creating file tombstone file with the updated state", "err", err)
+		return nil, err
+	}
+
+	if err = m.DeleteTombstoneFile(ctx, t.RequestID, t.State); err != nil {
+		level.Error(m.logger).Log("msg", "Created file with updated state but unable to delete previous state. Will retry next time tombstones are loaded", "err", err)
```
Comment (alvinlin123) [inline pkg/storage/tsdb/tombstones.go:247]: Would be nice to log the old and new state.

File: pkg/storage/tsdb/tombstones.go [hunk 14]
```diff
@@ -0,0 +1,360 @@
...
+}
+
+func (m *TombstoneManager) UpdateTombstoneState(ctx context.Context, t *Tombstone, newState BlockDeleteRequestState) (*Tombstone, error) {
+	// Create the new tombstone, and will delete the previous tombstone
+	newT := NewTombstone(t.UserID, t.RequestCreatedAt, time.Now().Unix()*1000, t.StartTime, t.EndTime, t.Selectors, t.RequestID, newState)
+	newT.Matchers = t.Matchers
+
+	err := m.WriteTombstone(ctx, newT)
+	if err != nil {
+		level.Error(m.logger).Log("msg", "error creating file tombstone file with the updated state", "err", err)
```
Comment (alvinlin123) [inline pkg/storage/tsdb/tombstones.go:242]: Would be nice to log the new state.

**Review by alvinlin123 — COMMENTED** [review]
I have not finished reading all the code yet, but I thought I have enough comment for this batch. I will continue the review at later time.

File: pkg/storage/tsdb/tombstones.go [hunk 15]
```diff
@@ -0,0 +1,360 @@
...
+
+	return false, nil
+}
+
+func (m *TombstoneManager) GetTombstoneByIDForUser(ctx context.Context, requestID string) (*Tombstone, error) {
+	found := []*Tombstone{}
+
+	for _, state := range AllDeletionStates {
+		filename := getTombstoneFileName(requestID, state)
+		exists, err := m.TombstoneExists(ctx, requestID, state)
```
Comment (alvinlin123) [inline pkg/storage/tsdb/tombstones.go:122]: I wonder if we really need this extra check, which causes extra network call (I believe the `m.TombstoneExists` translates to a S3 HEAD http call).

Would it be possible to just do `m.ReadTombstone` and differentiate if the returned `err` is NotFound or other errors, then handle them differently?

**Comment by ilangofman** [comment]
Sorry everyone, there has been no activity from me on this. I have been really busy with school and unfortunately haven't had time to work on this. I do plan to come back to this once I get a bit more time available, but if someone else wants to help out, that would be great too.

**Comment by alvinlin123** [comment]
@ilangofman  may start to work on this again soon :) so not staled.

File: pkg/chunk/purger/blocks_purger.go (hunk 10 above)
Comment (ilangofman) [reply pkg/chunk/purger/blocks_purger.go:193]: Yes, we can use the `deleteRequestCancelPeriod` to infer the processing state. Whenever the tombstone with statePending is older than `deleteRequestCancelPeriod`, then it is in processing state.

File: pkg/chunk/purger/blocks_purger.go (hunk 9 above)
Comment (ilangofman) [reply pkg/chunk/purger/blocks_purger.go:189]: Good point, thanks.

File: pkg/storage/tsdb/tombstones.go (hunk 12 above)
Comment (ilangofman) [reply pkg/storage/tsdb/tombstones.go:349]: Not sure why it was exported in the first place. I don't think the logic is required anywhere else. I made it private now.

File: pkg/chunk/purger/blocks_purger.go [hunk 16]
```diff
@@ -0,0 +1,234 @@
...
+
+	bufStart := make([]byte, 8)
+	binary.LittleEndian.PutUint64(bufStart, uint64(startTime))
+	bufEnd := make([]byte, 8)
+	binary.LittleEndian.PutUint64(bufEnd, uint64(startTime))
+
+	hash.Write(bufStart)
+	hash.Write(bufEnd)
+
+	sort.Strings(selectors)
```
[the remark this answers is not in the document]
Comment (ilangofman) [reply pkg/chunk/purger/blocks_purger.go]: I've updated it so it parses the matchers and then create the hash based on the parsed matchers. So any logically equivalent deletion requests should now result in the same hash. The example you provided should be the same hash. Also split up matchers like `match[]=process_start_time_seconds&match[]={job= "prometheus"}` would result in the same hash as the example you provided (Since a deletion request is the "AND" of all matchers provided).

File: pkg/chunk/purger/blocks_purger.go (hunk 7 above)
Comment (ilangofman) [reply pkg/chunk/purger/blocks_purger.go:161]: Makes sense, should be fixed now.

**Comment by CharlieTLe** [comment]
Hello @ilangofman, thank you for opening this PR.

There is a release in progress. As such, please rebase your CHANGELOG entry on top of the master branch and move the CHANGELOG entry to the top under `## master / unreleased`.

Thanks,
Charlie
